plugins/hls-signature-help-plugin/src/Ide/Plugin/SignatureHelp.hs

```diff
@@ -0,0 +1,208 @@
+            (Just $ InL argumentNumber)
+
+-- TODO(@linj) can type string be a multi-line string?
+mkArguments :: UInt -> Text -> [ParameterInformation]
+mkArguments offset functionType =
+    let separator = " -> "
+        separatorLength = fromIntegral $ T.length separator
+        splits = T.breakOnAll separator functionType
+        prefixes = fst <$> splits
+        prefixLengths = fmap (T.length >>> fromIntegral) prefixes
+        ranges =
+            [ ( if previousPrefixLength == 0 then 0 else previousPrefixLength + separatorLength,
+                currentPrefixLength
+              )
+            | (previousPrefixLength, currentPrefixLength) <- zip (0: prefixLengths) prefixLengths
+            ]
+     in [ ParameterInformation (InR range) Nothing
+        | range <- bimap (+offset) (+offset) <$> ranges
+        ]
+
+extractInfoFromSmallestContainingFunctionApplicationAst ::
+    Position -> HieASTs a -> (RealSrcSpan -> HieAST a -> Maybe b) -> [b]
+extractInfoFromSmallestContainingFunctionApplicationAst position hieAsts extractInfo =
+    M.elems $ flip M.mapMaybeWithKey (getAsts hieAsts) $ \hiePath hieAst ->
+        smallestContainingSatisfying (positionToSpan hiePath position) (nodeHasAnnotation ("HsApp", "HsExpr")) hieAst
+            >>= extractInfo (positionToSpan hiePath position)
+    where
+        positionToSpan hiePath position =
+            let loc = mkLoc hiePath position in mkRealSrcSpan loc loc
+        mkLoc (LexicalFastString hiePath) (Position line character) =
+            mkRealSrcLoc hiePath (fromIntegral line + 1) (fromIntegral character + 1)
+
+type Annotation = (FastStringCompat, FastStringCompat)
+
+nodeHasAnnotation :: Annotation -> HieAST a -> Bool
+nodeHasAnnotation annotation = sourceNodeInfo >>> maybe False (isAnnotationInNodeInfo annotation)
+
+-- TODO(@linj): the left most node may not be the function node. example: (if True then f else g) x
+getLeftMostNode :: HieAST a -> HieAST a
+getLeftMostNode thisNode =
+    case nodeChildren thisNode of
+        []           -> thisNode
+        leftChild: _ -> getLeftMostNode leftChild
+
+getNodeName :: RealSrcSpan -> HieAST a -> Maybe Name
+getNodeName _span hieAst =
+    if nodeHasAnnotation ("HsVar", "HsExpr") hieAst
+        then
+            case mapMaybe extractName $ M.keys $ M.filter isUse $ getSourceNodeIds hieAst of
+                [name] -> Just name -- TODO(@linj) will there be more than one name?
+                _      -> Nothing
+        else Nothing -- TODO(@linj) must function node be HsVar?
+    where
+        extractName = rightToMaybe
+
+-- TODO(@linj) share code with getNodeName
+getNodeType :: HieKind a -> RealSrcSpan -> HieAST a -> Maybe Text
+getNodeType (hieKind :: HieKind a) _span hieAst =
+    if nodeHasAnnotation ("HsVar", "HsExpr") hieAst
+       then
+           case M.elems $ M.filter isUse $ getSourceNodeIds hieAst of
+               [identifierDetails] -> identType identifierDetails >>= (prettyType >>> Just)
```

Comment:
I suggest not printing the type here and instead returning the type itself. That would mean that when you're doing `mkArguments` you can pattern match on the type itself, not on the string representation. Then you don't need to worry about e.g. whether it prints on multiple lines or whatever.

Reply:
See https://github.com/haskell/haskell-language-server/pull/4626#discussion_r2261133076